File: engine/src/flutter/lib/web_ui/lib/src/engine/skwasm/skwasm_impl/vertices.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// ignore_for_file: avoid_unused_constructor_parameters

import 'dart:ffi';
import 'dart:typed_data';

import 'package:ui/src/engine/skwasm/skwasm_impl.dart';
import 'package:ui/ui.dart' as ui;

class SkwasmVertices extends SkwasmObjectWrapper<RawVertices> implements ui.Vertices {
  factory SkwasmVertices(
    ui.VertexMode mode,
    List<ui.Offset> positions, {
    List<ui.Offset>? textureCoordinates,
    List<ui.Color>? colors,
    List<int>? indices,
  }) => withStackScope((StackScope scope) {
    final RawPointArray rawPositions = scope.convertPointArrayToNative(positions);
    final RawPointArray rawTextureCoordinates =
        textureCoordinates != null ? scope.convertPointArrayToNative(textureCoordinates) : nullptr;
    final RawColorArray rawColors =
        colors != null ? scope.convertColorArrayToNative(colors) : nullptr;
    final Pointer<Uint16> rawIndices =
        indices != null ? scope.convertIntsToUint16Native(indices) : nullptr;
    final int indexCount = indices != null ? indices.length : 0;
    return SkwasmVertices._(
      verticesCreate(
        mode.index,
        positions.length,
        rawPositions,
        rawTextureCoordinates,
        rawColors,
        indexCount,
        rawIndices,
      ),
    );
  });

  factory SkwasmVertices.raw(
    ui.VertexMode mode,
    Float32List positions, {
    Float32List? textureCoordinates,
    Int32List? colors,
    Uint16List? indices,
  }) => withStackScope((StackScope scope) {
    final RawPointArray rawPositions = scope.convertDoublesToNative(positions);
    final RawPointArray rawTextureCoordinates =
        textureCoordinates != null ? scope.convertDoublesToNative(textureCoordinates) : nullptr;
    final RawColorArray rawColors =
        colors != null ? scope.convertIntsToUint32Native(colors) : nullptr;
    final Pointer<Uint16> rawIndices =
        indices != null ? scope.convertIntsToUint16Native(indices) : nullptr;
    final int indexCount = indices != null ? indices.length : 0;
    return SkwasmVertices._(
      verticesCreate(
        mode.index,
        positions.length ~/ 2,
        rawPositions,
        rawTextureCoordinates,
        rawColors,
        indexCount,
        rawIndices,
      ),
    );
  });

  SkwasmVertices._(VerticesHandle handle) : super(handle, _registry);

  static final SkwasmFinalizationRegistry<RawVertices> _registry =
      SkwasmFinalizationRegistry<RawVertices>(verticesDispose);
}

File: engine/src/flutter/skwasm/vertices.cc
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <vector>

#include "flutter/display_list/dl_vertices.h"
#include "flutter/skwasm/export.h"
#include "flutter/skwasm/helpers.h"
#include "flutter/skwasm/live_objects.h"

SKWASM_EXPORT Skwasm::sp_wrapper<flutter::DlVertices>* vertices_create(
    flutter::DlVertexMode vertex_mode,
    int vertex_count,
    flutter::DlPoint* positions,
    flutter::DlPoint* texture_coordinates,
    uint32_t* colors,
    int index_count,
    uint16_t* indices) {
  Skwasm::live_vertices_count++;
  std::vector<flutter::DlColor> dl_colors;
  flutter::DlColor* dl_color_pointer = nullptr;
  if (colors != nullptr) {
    dl_colors.resize(vertex_count);
    for (int i = 0; i < vertex_count; i++) {
      dl_colors[i] = flutter::DlColor(colors[i]);
    }
    dl_color_pointer = dl_colors.data();
  }
  return new Skwasm::sp_wrapper<flutter::DlVertices>(flutter::DlVertices::Make(
      vertex_mode, vertex_count, positions, texture_coordinates,
      dl_color_pointer, index_count, indices));
}

SKWASM_EXPORT void vertices_dispose(
    Skwasm::sp_wrapper<flutter::DlVertices>* vertices) {
  Skwasm::live_vertices_count--;
  delete vertices;
}
